scylla/tests/integration/tablets.rs

```diff
@@ -0,0 +1,586 @@
+        },
+    )
+    .await;
+    match res {
+        Ok(()) => (),
+        Err(ProxyError::Worker(WorkerError::DriverDisconnected(_))) => (),
+        Err(err) => panic!("{}", err),
+    }
+}
+
+/// This test verifies that Scylla never sends tablet info when receiving unprepared
+/// query - as it would make no sens to send it (driver can't do token awareness
+/// for unprepared queries).
+///
+/// The test sends a query to each shard of every node and verifies that no
+/// tablet info was sent in response.
+#[tokio::test]
+#[ntest::timeout(30000)]
+async fn test_tablet_feedback_not_sent_for_unprepared_queries() {
+    setup_tracing();
+    const TABLET_COUNT: usize = 16;
+
+    let res = test_with_3_node_cluster(
+        ShardAwareness::QueryNode,
+        |proxy_uris, translation_map, mut running_proxy| async move {
+            let session = scylla::SessionBuilder::new()
+                .known_node(proxy_uris[0].as_str())
+                .address_translator(Arc::new(translation_map))
+                .build()
+                .await
+                .unwrap();
+
+            if !scylla::test_utils::scylla_supports_tablets(&session).await {
+                tracing::warn!("Skipping test because this Scylla version doesn't support tablets");
+                return running_proxy;
+            }
+
+            let ks = unique_keyspace_name();
+
+            /* Prepare schema */
+            prepare_schema(&session, &ks, "t", TABLET_COUNT).await;
+
+            let (feedback_txs, mut feedback_rxs): (Vec<_>, Vec<_>) = (0..3)
+                .map(|_| mpsc::unbounded_channel::<(ResponseFrame, Option<TargetShard>)>())
+                .unzip();
+            for (i, tx) in feedback_txs.iter().cloned().enumerate() {
+                running_proxy.running_nodes[i].change_response_rules(Some(vec![ResponseRule(
+                    Condition::ResponseOpcode(ResponseOpcode::Result)
+                        .and(Condition::not(Condition::ConnectionRegisteredAnyEvent)),
+                    ResponseReaction::noop().with_feedback_when_performed(tx),
+                )]));
+            }
+
+            // I expect Scylla to not send feedback for unprepared queries,
+            // as such queries cannot be token-aware anyway
+            send_unprepared_query_everywhere(
+                &session,
+                session.get_cluster_data().as_ref(),
+                &Query::new(format!("INSERT INTO {ks}.t (a, b, c) VALUES (1, 1, 'abc')")),
+            )
+            .await
+            .unwrap();
+
+            let feedbacks: usize = feedback_rxs.iter_mut().map(count_tablet_feedbacks).sum();
+            assert!(feedbacks == 0);
+
+            running_proxy
+        },
+    )
+    .await;
+    match res {
+        Ok(()) => (),
+        Err(ProxyError::Worker(WorkerError::DriverDisconnected(_))) => (),
+        Err(err) => panic!("{}", err),
+    }
+}
+
+/// Verifies that LWT optimization (sending LWT queries to a single replica always)
```

Comment:
I'm unsure about this definition of the LWT optimization. I believe it could be better described by _sending LWT queries to the replicas in a fixed order_; there can be more replicas in the query plan than just a single one.